File: packages/api/src/models/report.ts
import { ObjectId } from 'mongodb';

/**
 * Report document schema with TTL index
 * Requirements: 8.1, 8.3, 10.1, 11.1
 */

export interface ReportMetadata {
  generatedAt: Date;
  executionTime: number;
  dataSourcesUsed: string[];
  version: string;
}

export interface ReportSection {
  title: string;
  content: string;
  order: number;
  visualizationIds?: string[];
}

export interface Report {
  title: string;
  executiveSummary: string;
  sections: ReportSection[];
  metadata: ReportMetadata;
}

export interface Anomaly {
  metric: string;
  timestamp: number;
  value: number;
  expectedValue: number;
  deviationScore: number;
  severity: 'low' | 'medium' | 'high';
  context?: string;
}

export interface Correlation {
  metric1: string;
  metric2: string;
  coefficient: number;
  pValue: number;
  significance: boolean;
  lag?: number;
}

export interface Pattern {
  type: string;
  metric: string;
  description: string;
  confidence: number;
  timeRange: {
    start: number;
    end: number;
  };
}

export interface Statistics {
  mean: number;
  median: number;
  stdDev: number;
  min: number;
  max: number;
  count: number;
}

export interface AnalysisResponse {
  correlations: Correlation[];
  anomalies: Anomaly[];
  patterns: Pattern[];
  statistics: Record<string, Statistics>;
}

export interface Evidence {
  source: string;
  value: number | string;
  timestamp: number;
}

export interface VerifiedClaim {
  id: string;
  statement: string;
  metric: string;
  value: number | string;
  verified: boolean;
  confidence: number;
  sources: string[];
  evidence: Evidence[];
}

export interface DataConflict {
  claimId: string;
  sources: {
    source: string;
    value: number | string;
    difference: number;
  }[];
  resolution?: string;
}

export interface AuditEntry {
  timestamp: Date;
  action: string;
  claimId: string;
  result: 'verified' | 'failed' | 'conflict';
  details: Record<string, any>;
}

export interface FactCheckResponse {
  verifiedClaims: VerifiedClaim[];
  conflicts: DataConflict[];
  auditTrail: AuditEntry[];
  overallConfidence: number;
}

export interface ReportDocument {
  _id?: ObjectId;
  reportId: string;
  userId: string;
  queryId: string;
  query: string;
  createdAt: Date;
  report: Report;
  analysisResults: AnalysisResponse;
  factCheckResults?: FactCheckResponse;
  visualizationUrls: string[];
  expiresAt: Date;
}

export const REPORT_COLLECTION = 'reports';

/**
 * MongoDB indexes for report collection
 * Includes TTL index for automatic expiration
 */
export const REPORT_INDEXES = [
  {
    key: { reportId: 1 },
    unique: true,
    name: 'idx_reportId',
  },
  {
    key: { userId: 1 },
    name: 'idx_userId',
  },
  {
    key: { queryId: 1 },
    name: 'idx_queryId',
  },
  {
    key: { userId: 1, createdAt: -1 },
    name: 'idx_userId_createdAt',
  },
  {
    key: { expiresAt: 1 },
    expireAfterSeconds: 0,
    name: 'idx_ttl',
  },
];

File: packages/agents/src/models/report.py
"""Report document models for MongoDB"""
from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field


class ReportMetadata(BaseModel):
    """Report metadata"""
    generated_at: datetime
    execution_time: float
    data_sources_used: List[str]
    version: str = "1.0"


class ReportSection(BaseModel):
    """Report section"""
    title: str
    content: str
    order: int
    visualization_ids: Optional[List[str]] = None


class Report(BaseModel):
    """Report content"""
    title: str
    executive_summary: str
    sections: List[ReportSection]
    metadata: ReportMetadata


class Anomaly(BaseModel):
    """Anomaly detection result"""
    metric: str
    timestamp: int
    value: float
    expected_value: float
    deviation_score: float
    severity: str
    context: Optional[str] = None


class Correlation(BaseModel):
    """Correlation analysis result"""
    metric1: str
    metric2: str
    coefficient: float
    p_value: float
    significance: bool
    lag: Optional[int] = None


class Pattern(BaseModel):
    """Pattern recognition result"""
    type: str
    metric: str
    description: str
    confidence: float
    time_range: Dict[str, int]


class Statistics(BaseModel):
    """Statistical summary"""
    mean: float
    median: float
    std_dev: float
    min: float
    max: float
    count: int


class AnalysisResponse(BaseModel):
    """Analysis results"""
    correlations: List[Correlation] = Field(default_factory=list)
    anomalies: List[Anomaly] = Field(default_factory=list)
    patterns: List[Pattern] = Field(default_factory=list)
    statistics: Dict[str, Statistics] = Field(default_factory=dict)


class Evidence(BaseModel):
    """Verification evidence"""
    source: str
    value: Any
    timestamp: int


class VerifiedClaim(BaseModel):
    """Verified claim"""
    id: str
    statement: str
    metric: str
    value: Any
    verified: bool
    confidence: float
    sources: List[str]
    evidence: List[Evidence]


class DataConflict(BaseModel):
    """Data conflict between sources"""
    claim_id: str
    sources: List[Dict[str, Any]]
    resolution: Optional[str] = None


class AuditEntry(BaseModel):
    """Audit trail entry"""
    timestamp: datetime
    action: str
    claim_id: str
    result: str
    details: Dict[str, Any]


class FactCheckResponse(BaseModel):
    """Fact check results"""
    verified_claims: List[VerifiedClaim] = Field(default_factory=list)
    conflicts: List[DataConflict] = Field(default_factory=list)
    audit_trail: List[AuditEntry] = Field(default_factory=list)
    overall_confidence: float


class ReportDocument(BaseModel):
    """Report document"""
    report_id: str
    user_id: str
    query_id: str
    query: str
    created_at: datetime = Field(default_factory=datetime.utcnow)
    report: Report
    analysis_results: AnalysisResponse
    fact_check_results: Optional[FactCheckResponse] = None
    visualization_urls: List[str] = Field(default_factory=list)
    expires_at: datetime

    class Config:
        """Pydantic configuration"""
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }


# Collection name
REPORT_COLLECTION = "reports"

# Index specifications
REPORT_INDEXES = [
    {"keys": [("report_id", 1)], "unique": True, "name": "idx_reportId"},
    {"keys": [("user_id", 1)], "name": "idx_userId"},
    {"keys": [("query_id", 1)], "name": "idx_queryId"},
    {"keys": [("user_id", 1), ("created_at", -1)], "name": "idx_userId_createdAt"},
    {
        "keys": [("expires_at", 1)],
        "expireAfterSeconds": 0,
        "name": "idx_ttl"
    },
]
